proj/hps_accel/gateware/gen2/sysarray.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Gateware for the accelerator."""
+
+from nmigen import Signal, signed, unsigned
+from nmigen_cfu import SimpleElaboratable
+from .constants import Constants
+from .macc import MaccBlock
+
+
+class SystolicArray(SimpleElaboratable):
+    """A systolic array of multiply-accumulators.
+
+    Parameters
+    ----------
+
+    a_size: int
+      The number of A words processed concurrently
+
+    b_size: int
+      The number of B words processed concurrently
+
+    n: int
+        The number of multipliers. Four is the usual value.
+
+    a_shape: Shape
+        The shape of each the value packed into the A input. If the A
+        input is used for Conv2D activation values, signed(9) or
+        unsigned(8) might be appropriate values.
+
+    b_shape: Shape
+        The shape of each the value packed into the B input. If the B
+        input is used for Conv2D filter values, signed(8) would be the
+        appropriate shape.
+
+    accumulator_shape: Shape
+        The shape of the accumulator. The accumulator should have
+        enough precision that it does not overflow under normal
+        operation. signed(32) is usually large enough, but a narrower
+        accumulator may be sufficient, and will make more efficient
+        use of the FPGA fabric.
+
+    Attributes
+    ----------
+
+    a_inputs: [Signal(n * a_shape.width)], in
+      List of words forming the A inputs
+
+    b_inputs: [Signal(32)], in
```

Comment:
I guess this is actually a list of `Signal(n * b_shape.width)` which in our case is `Signal(32)`, right?

Reply:
@alanvgreen why the difference between how `a_inputs` and `b_inputs` are declared?

Reply:
No good reason. Fixed.

Reply:
@alanvgreen did you push your change, or am I misunderstanding something?

Reply:
fixed now